### PR TITLE
Sharable Kedro-Viz UI for AWS, GCP and Azure

### PR DESCRIPTION
**Pull request by jitu5**
## Description

Resolves [#1674](https://github.com/kedro-org/kedro-viz/issues/1674) & [#1675](https://github.com/kedro-org/kedro-viz/issues/1675)

Sharable URL modal UI updated with below changes

Screenshot 2024-01-25 at 3 11 02 p m
Screenshot 2024-01-25 at 3 10 50 p m

## Development notes

* [shareable-url-modal](https://github.com/kedro-org/kedro-viz/pull/1718/files#diff-6706d5ddbf6aa262c3b985c7d9d3274a7589da883cad87ef27c2c8c21c7740fc) component updated as per new design
* [shareable-url-modal](https://github.com/kedro-org/kedro-viz/pull/1718/files#diff-6706d5ddbf6aa262c3b985c7d9d3274a7589da883cad87ef27c2c8c21c7740fc) e2e cypress test updated
* In [metadata.js](https://github.com/kedro-org/kedro-viz/pull/1718/files#diff-75f06a5a17b3eeb6bde03d6cfffc97b833de28b38c08a36928d432d2754c8492) and [settings-modal.js](https://github.com/kedro-org/kedro-viz/pull/1718/files#diff-d788d55b37bec9ab529b7cbfa518ce2c40fb56d3abb265ac1890830b644cedc9) version info and experiment tracking link now conditionally hidden 
* `<Input>` component updated to accommodate `input` and `textarea` type
* `<Dropdown>` component updated to accommodate “placehoder” text
* `<Modal>` component updated to make `title` conditional

## QA notes

**Note:** This will be merged into the branch `feature/shareableviz-extended-support`. 
Deploying to AWS, GCP and Azure will only works from UI after [1708](https://github.com/kedro-org/kedro-viz/pull/1708) & [1711](https://github.com/kedro-org/kedro-viz/pull/1711) get merged into the branch `feature/shareableviz-extended-support`

## Checklist

- [X] Read the [contributing](/CONTRIBUTING.md) guidelines
- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [X] Added tests to cover my changes

### CODE REVIEW

File: src/components/experiment-tracking/run-details-modal/run-details-modal.js [hunk 1]
```diff
@@ -87,6 +87,7 @@ const RunDetailsModal = ({
             defaultValue={valuesToUpdate.title}
             onChange={(value) => onChange('title', value)}
             resetValueTrigger={visible}
+            type="textarea"
```
Comment (rashidakanchwala) [inline src/components/experiment-tracking/run-details-modal/run-details-modal.js:90]: hi, what is this for ?

Comment (jitu5) [reply src/components/experiment-tracking/run-details-modal/run-details-modal.js:90]: `<Input>` component uses <textarea> internally but its not suitable for input fields in sharable-viz modal so I modified `<Input>` component to accept both the type `textarea` & `input` 
See here [Input component](https://github.com/kedro-org/kedro-viz/pull/1718/files#diff-1264eb0cfb0afbaed97f98d83a95263a0e1434850cae86844952eaeef7caa24f)

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 2]
```diff
@@ -125,83 +128,157 @@ const ShareableUrlModal = ({ onToggleModal, visible }) => {
...
     setResponseError(null);
     setIsLoading(false);
     setResponseUrl(null);
     setIsLinkSettingsClick(false);
     setInputValues({});
     setIsFormDirty({
       bucket_name: false,
-      region: false,
+      platform: false,
+      endpoint: false,
```
Comment (ravi-kumar-pilla) [inline src/components/shareable-url-modal/shareable-url-modal.js:142]: same as above (has_endpoint)

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 3]
```diff
@@ -125,83 +128,157 @@ const ShareableUrlModal = ({ onToggleModal, visible }) => {
...
+      return deploymentState === 'success'
+        ? 'Kedro-Viz Published and Hosted'
+        : 'Publish and Share Kedro-Viz';
+    }
+
+    return modalMessages(deploymentState, compatibilityData.package_version);
+  };
+
+  const handleResponseUrl = () => {
+    // If the URL does not start with http:// or https://, append http:// to avoid relative path issue
```
Comment (ravi-kumar-pilla) [inline src/components/shareable-url-modal/shareable-url-modal.js:166]: appending **http** will not work for Azure. By default AzureBlobStorage serves content on **https**
`Secure transfer required is enabled by default`

image

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 4]
```diff
@@ -125,83 +128,157 @@ const ShareableUrlModal = ({ onToggleModal, visible }) => {
...
+    if (!/^https?:\/\//.test(responseUrl)) {
+      const url = 'http://' + responseUrl;
+      return url;
+    }
+    return responseUrl;
+  };
```
Comment (ravi-kumar-pilla) [inline src/components/shareable-url-modal/shareable-url-modal.js:172]: I am not sure how to handle this if the user enters a wrong endpoint `url` . There can be many cases of url not entered properly. We should document this in our docs and avoid handling all cases. I think you did this to handle gcp load balancer usecase @jitu5 ?

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 5]
```diff
@@ -125,83 +128,157 @@ const ShareableUrlModal = ({ onToggleModal, visible }) => {
...
+      endpoint: false,
     }); /* eslint-disable camelcase */
   };
 
+  /**
+   * Returns the modal heading based on the given type and deployment state.
+   * @param {string} type - The type of the modal heading.
+   * @returns {string|null} The modal heading text or null if deployment state is 'default'.
+   */
+  const getModalHeading = (type) => {
```
Comment (ravi-kumar-pilla) [inline src/components/shareable-url-modal/shareable-url-modal.js:151]: [Nit] `getModalHeading` is a bit confusing to me. Can we rename to `getDeploymentStateTitle` (not sure of this either)

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 6]
```diff
@@ -125,83 +128,157 @@ const ShareableUrlModal = ({ onToggleModal, visible }) => {
...
+          <div className="shareable-url-modal__content-form-wrapper">
+            <div className="shareable-url-modal__content-wrapper">
+              <div className="shareable-url-modal__content-title">
+                Publish and Share Kedro-Viz
+              </div>
+              <p className="shareable-url-modal__content-description shareable-url-modal__paregraph-divider">
+                Prerequisite: Deploying and hosting Kedro-Viz requires access
+                keys or user credentials, depending on the chosen service
+                provider. To use this feature, please add your access keys or
+                credentials as environment variables in your project. More
```
Comment (ravi-kumar-pilla) [inline src/components/shareable-url-modal/shareable-url-modal.js:198]: [Nit] credentials as environment variables in your **kedro** project ?

File: src/components/ui/input/index.js [hunk 7]
```diff
@@ -45,16 +47,32 @@ const Input = ({
...
+        placeholder={placeholder}
+        ref={ref}
+        value={value}
+        data-test={dataTest}
+      />
+    );
+
+  return (
+    <>
+      {inputElement}
```
Comment (ravi-kumar-pilla) [inline src/components/ui/input/index.js:75]: something not related to the PR but we have (https://github.com/kedro-org/kedro-viz/blob/main/src/styles/_extends.scss) which has `cursor: pointer` for input text boxes. For something where user needs to type in, can we remove this default ? @jitu5 @tynandebold @rashidakanchwala

**Comment by ravi-kumar-pilla** [comment]
Hi Team,

We also need to handle the case where the credentials are configured correctly but the bucket_name is incorrect. Below is the output and error type as of now -

**AWS**
BE Error - `FileNotFoundError: The specified bucket does not exist ` 

FE -

Screenshot 2024-02-20 at 3 38 19 PM

**Azure**
BE Error - `ServiceRequestError: Cannot connect to host somek.blob.core.windows.net:443 ssl:default [nodename nor servname provided, or not known]`

FE -

Screenshot 2024-02-20 at 3 37 45 PM

**GCP**
BE Error - `FileNotFoundError: https://storage.googleapis.com/upload/storage/v1/b/somek/o`

FE -

Publish and Share Kedro-Viz


Should we create a separate ticket and handle these scenarios by passing a common error message like - `The specified bucket does not exist` when the BE error is either `FileNotFoundError` or `ServiceRequestError` ? @rashidakanchwala 

Thank you

**Comment by rashidakanchwala** [comment]
Hi Ravi, great point, let's create a seperate ticket to handle error mesages as we will have to think through this a bit more. Once Steph is back, we can map out different success/failure journeys.

File: src/components/ui/input/index.js (hunk 7 above)
Comment (rashidakanchwala) [reply src/components/ui/input/index.js:75]: good shout .. can we create an issue and add it as quick win.

File: src/components/shareable-url-modal/shareable-url-modal.js (hunk 4 above)
Comment (jitu5) [reply src/components/shareable-url-modal/shareable-url-modal.js:172]: In the case of GCP with a load balancer, the endpoint will most likely be an IP address, such as `32.206.61.13:80`. If a user pastes this IP address into the endpoint input field, the clickable link on the success page becomes relative to the current address, like `https://example.com/32.206.61.13:80`. To address this, I have appended 'http://' to the link.

Should this handling be specific to GCP?

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 8]
```diff
@@ -35,8 +37,9 @@ const ShareableUrlModal = ({ onToggleModal, visible }) => {
   const [inputValues, setInputValues] = useState({});
   const [isFormDirty, setIsFormDirty] = useState({
     /* eslint-disable camelcase */
-    bucket_name: false,
-    region: false,
+    has_bucket_name: false,
+    has_platform: false,
+    has_endpoint: false,
```
Comment (ravi-kumar-pilla) [inline src/components/shareable-url-modal/shareable-url-modal.js:42]: my bad, I was giving an example but if we can use camelCase we should do that like (hasBucketName)

**Comment by ravi-kumar-pilla** [comment]
> Hi Ravi, I also agree the above errors are quite vague. As a short term solution -- let's have a common - 'The specified bucket does not exist' for FileNotFound or ServiceRequestError.

Sure will add it in the current branch.

> But let's also create a seperate ticket to handle error mesages as we will have to think through this a bit more. Once Steph is back, we can map out different success/failure journeys.

Sure, will do that

**Comment by ravi-kumar-pilla** [comment]
Hi Team,

This might be a design question -> when we publish successfully and click on **link settings** -> The publish button text displays as **Republish**. This is great but if we change the hosting platform or change the bucket_name, will the new publishing be a re-publish or just publish ?

image

I feel we should stick to the text as **Publish** in both the cases. Thank you

**Comment by rashidakanchwala** [comment]
> Hi Team,
> 
> This might be a design question -> when we publish successfully and click on **link settings** -> The publish button text displays as **Republish**. This is great but if we change the hosting platform or change the bucket_name, will the new publishing be a re-publish or just publish ?
> 
> image
> I feel we should stick to the text as **Publish** in both the cases. Thank you

You're correct. I also found 'Republish' a bit confusing. @stephkaiser's original design involved guiding users through multiple selections. Initially, you choose the 'cloud provider,' then click 'Next' to proceed to the form where you enter details. However, upon returning to 'Link Settings,' it directs you straight to the form page without an option to switch 'Cloud Provider.' We diverged from Steph's original design because it assumed each Cloud Provider would have different input fields. However, since they're now the same, the two-step process feels unnecessarily lengthy.

When Steph returns, perhaps we can refine this journey. In the meantime, I propose that when the user clicks 'Link Settings', the dropdown for selecting the cloud provider should be disabled. This adjustment would make 'Republish' more intuitive. Otherwise, we should revert to using 'Publish.'

**Comment by jitu5** [comment]
@rashidakanchwala @ravi-kumar-pilla For simplification I am keeping the text as Publish in both the cases.

**Comment by rashidakanchwala** [comment]
Amazing, LGTM. thanks Jitendra!! <3

**Review by rashidakanchwala — APPROVED** [review]

**Pull request merged**